docs/internal/expressions.md

```diff
@@ -0,0 +1,611 @@
+For example the type name (`Foo::method()`), the first argument of a pre-fix operator (`foo.method()` or `method(foo)`), or the left hand side of an infix operator (`foo method bar`).
+
+Example:
+```ts
+// Defines the type of `x`
+let x : U32 & imm = 42;
+
+// Infers the type of `y` and `10` from the type of `x`
+// Checks that `U32::+(U32, U32)` exists
+// Calls U32::+(x, 10)
+let y = x + 10;
+
+// The above is identical to ones below
+let a = x.+(10);
+let b = U32::+(x, 10);
+```
+
+The Verona standard library will implement all operations for all appropriate types, including all arithmetic for all numeric types and string handling for all character and string types.
+Users can implement similar calls to specialised type (ex. `MyInt::+(MyInt, I32)`).
+
+There is no precedence in Verona, so operators must be wrapped in parenthesis to convey the right meaning.
+This means that, unlike other languages, `(a + b * c)` is a syntax error, while either `((a + b) * c)` or `(a + (b * c))` are expected.
+
+_Note: Can we avoid forcing parenthesis on DSLs?_
+
+## Arithmetic
+
+As detailed above, arithmetic in Verona is implemented as function calls on numeric types and operands.
+This means developers can extend the functionality of arithmetic to the existing types or their own new types naturally.
+But it also means that every addition or subtraction could end up as call on objects, which could potentially be many orders of magnitude slower than standard hardware operations.
+
+But Verona's arithmetic is often as fast as calling the right hardware instructions.
+
+The reasons are:
+1. As explained in the types document, numeric types are singleton classes, ie. they don't have fields, only methods.
+   The compiler treats them differently and uses their machine representation instead of a pointer to an allocated memory.
+2. Computation often uses concrete types (ex. `U32` or `F64`) instead of interface types (ex. `IntegralType`) or type unions (ex. `(U32 | F64`)).
+   The compiler can only match concrete types to their machine equivalent.
+   It can try to infer, match and separate concrete types from their declared collection types, but that's an optimisation, not a guarantee.
+3. Direct code generation of the hardware instructions for arithmetic functions.
+   Those functions are often short and simple and are always inlined by the compiler.
+
+This means `(a + b)` in Verona usually completely bypasses function calls and wrappers and just call a single `add` instruction.
+
+### Semantics
+
+There is no undefined behaviour in Verona.
+There are no traps or run-time exceptions from arithmetic in Verona.
+
+Verona's integer arithmetic has wrap-around semantics as defined behaviour.
+Adding unsigned values past their maximum values wraps around to zero, subtracting signed values past their minimum negative values wraps around to positive, etc.
+
+Integer division by zero is zero.
```

Comment:
Is there an efficient implementation of this?  I can see it for `x / 0 == 1` by doing `x / n = x / ((n == 0) ? x : n)`, which is a single extra instruction (maybe two) and is amenable to vectorisation, but I don't know how you get the 0 behaviour without a branch.